File: mvc/src/main/java/com/example/icecream/AssessmentRecordActivity.java
package com.example.icecream;

import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.widget.Toast;

import com.example.icecream.databinding.ActivityAssessmentRecordBinding;
import com.example.icecream.di.assessment_record.AssessmentRecordSubcomponent;
import com.example.icecream.di.assessment_record.ControllerModule;
import com.example.icecream.di.assessment_record.ModelModule;
import com.example.icecream.model.AssessmentRecord;
import com.example.icecream.model.AssessmentRecordsManager;
import com.example.icecream.network.NetworkStateIndicator;

import javax.inject.Inject;

public class AssessmentRecordActivity extends AppCompatActivity {

    private final String KEY_RECORDS = "key_records";

    @Inject
    AssessmentRecordsManager recordsManager;

    @Inject
    AssessmentRecord record;

    @Inject
    NetworkStateIndicator networkStateIndicator;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        ActivityAssessmentRecordBinding binding =
                DataBindingUtil.setContentView(this, R.layout.activity_assessment_record);
        initialize(binding, savedInstanceState);
    }

    @Override
    protected void onResume() {
        super.onResume();
        recordsManager.loadRecords();
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        if (isChangingConfigurations()) {
            Bundle savedState = recordsManager.getSavedState();
            outState.putBundle(KEY_RECORDS, savedState);
        }
    }

    public void showToast(String message) {
        Toast.makeText(this, message, Toast.LENGTH_LONG).show();
    }

    private void initialize(ActivityAssessmentRecordBinding binding, Bundle savedInstanceState) {
        Bundle modelState = null;
        if (savedInstanceState != null) {
            modelState = savedInstanceState.getBundle(KEY_RECORDS);
        }
        AssessmentRecordSubcomponent subcomponent = IceCreamApplication.component
                .plus(new ControllerModule(), new ModelModule(modelState));

        subcomponent.inject(this);

        binding.setRecordsManager(recordsManager);
        binding.setRecord(record);
        binding.setActualController(subcomponent.actualFieldController());
        binding.setStationIdsController(subcomponent.stationIdsSpinnerController());
        binding.setSendButtonController(subcomponent.sendButtonController());
        binding.setRefreshController(subcomponent.refreshLayoutController());
        binding.setNetwork(networkStateIndicator);
    }
}

File: mvvm/src/main/java/com/example/icecream/AssessmentRecordActivity.kt
package com.example.icecream

import android.databinding.DataBindingUtil
import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import com.example.icecream.databinding.ActivityAssessmentRecordBinding
import com.example.icecream.di.assessment_record.ModelModule
import com.example.icecream.di.assessment_record.ViewModelModule
import com.example.icecream.view_model.AssessmentRecordViewModel
import com.example.icecream.wrapper.LifecycleEventsConsumer
import javax.inject.Inject

class AssessmentRecordActivity : AppCompatActivity() {

    @Inject lateinit var viewModel: AssessmentRecordViewModel

    @Inject lateinit var lifecycleEventsConsumer: LifecycleEventsConsumer

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val binding: ActivityAssessmentRecordBinding =
                DataBindingUtil.setContentView(this, R.layout.activity_assessment_record)

        val savedState = savedInstanceState?.getBundle(KEY_STATE)

        IceCreamApplication.component
                .plus(ModelModule(savedState), ViewModelModule(this))
                .inject(this)

        binding.viewModel = viewModel

        lifecycleEventsConsumer.onCreate()
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putBundle(KEY_STATE, viewModel.saveState())
    }

    override fun onDestroy() {
        lifecycleEventsConsumer.onDestroy()
        super.onDestroy()
    }
}

private const val KEY_STATE = "key_state"